File: be/src/main/java/sg/toss_sg/repositories/account/AccountRepositoryImpl.java
package sg.toss_sg.repositories.account;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Timestamp;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

import org.springframework.stereotype.Repository;

import sg.toss_sg.configs.DatabaseConnectionPool;
import sg.toss_sg.entities.Account;
import sg.toss_sg.entities.Bank;
import sg.toss_sg.entities.User;
import sg.toss_sg.entities.utils.AccountType;
import sg.toss_sg.models.account.AccountWithTransactionHistory;
import sg.toss_sg.models.account.BriefAccount;
import sg.toss_sg.repositories.transactionHistory.TransactionHistoryRepository;
import sg.toss_sg.repositories.utils.AccountQueryStore;

@Repository
public class AccountRepositoryImpl implements AccountRepository {

    DatabaseConnectionPool databaseConnectionPool;

    TransactionHistoryRepository transactionHistoryRepository;

    private final String SAVE = AccountQueryStore.SAVE_ACCOUNT;

    private final String SAVE_WITH_ID = AccountQueryStore.SAVE_ACCOUNT_WITH_ID;

    private final String FIND_BY_ID = AccountQueryStore.FIND_ACCOUNT_BY_ID;

    private final String DELETE_ALL = AccountQueryStore.DELETE_ALL_ACCOUNTS;

    private final String FIND_ACCOUNT_OF_USER = AccountQueryStore.FIND_BRIEF_ACCOUNT_OF_USER;

    private final String FIND_ACCOUNT_WITH_TRANSACTION_HISTORY_OF_USER = AccountQueryStore.FIND_ACCOUNT_WITH_TRANSACTION_HISTORY_OF_USER;

    public AccountRepositoryImpl(DatabaseConnectionPool databaseConnectionPool,
            TransactionHistoryRepository transactionHistoryRepository) {
        this.databaseConnectionPool = databaseConnectionPool;
        this.transactionHistoryRepository = transactionHistoryRepository;
    }

    @Override
    public Account save(Account entity) {
        Connection connection = databaseConnectionPool.getConnection();
        boolean hasId = entity.getId() != null;
        String queryString = hasId ? SAVE_WITH_ID : SAVE;
        int parameterStart = hasId ? 1 : 0;
        try (PreparedStatement pstm = connection.prepareStatement(queryString)) {
            if (hasId) {
                pstm.setLong(parameterStart, entity.getId());
            }
            pstm.setString(parameterStart + 1, entity.getAccountNumber());
            pstm.setInt(parameterStart + 2, entity.getBank().getId());
            pstm.setInt(parameterStart + 3, entity.getOwner().getId());
            pstm.setDouble(parameterStart + 4, entity.getBalance());
            pstm.setString(parameterStart + 5, entity.getAccountName());
            pstm.setString(parameterStart + 6, entity.getAccountType().toString());
            pstm.setTimestamp(parameterStart + 7, Timestamp.valueOf(entity.getLastUpdated()));

            pstm.executeUpdate();
            connection.close();
            return entity;
        } catch (SQLException e) {
            e.printStackTrace();
            return null;
        } finally {
            try {
                connection.close();
            } catch (SQLException e1) {
                e1.printStackTrace();
            }
        }
    }

    @Override
    public Optional<Account> findById(Long id) {
        if (id == null) {
            return Optional.empty();
        }

        Connection connection = databaseConnectionPool.getConnection();

        try (PreparedStatement pstm = connection.prepareStatement(FIND_BY_ID)) {
            pstm.setLong(1, id);

            ResultSet resultSet = pstm.executeQuery();
            if (resultSet.isLast()) {
                return Optional.empty();
            }
            if (resultSet.next()) {
                Account account = Account.builder()
                        .id(resultSet.getLong("id"))
                        .accountNumber(resultSet.getString("account_number"))
                        .balance(resultSet.getDouble("balance"))
                        .accountName(resultSet.getString("account_name"))
                        .accountType(AccountType.valueOf(resultSet.getString("account_type")))
                        .lastUpdated(resultSet.getTimestamp("last_updated").toLocalDateTime())
                        .bank(Bank.builder()
                                .id(resultSet.getInt("bank_id"))
                                .name(resultSet.getString("bank_name"))
                                .bankCode(resultSet.getString("bank_code"))
                                .build())
                        .owner(User.builder()
                                .id(resultSet.getInt("user_id"))
                                .name(resultSet.getString("name"))
                                .email(resultSet.getString("email"))
                                .identificationNumber(resultSet.getString("identification_number"))
                                .phoneNumber(resultSet.getString("phone_number"))
                                .dateOfBirth(resultSet.getDate("date_of_birth").toLocalDate())
                                .settingJson(resultSet.getString("setting_json"))
                                .build())
                        .build();
                return Optional.of(account);
            }
            return Optional.empty();
        } catch (SQLException e) {
            e.printStackTrace();
            return null;
        } finally {
            try {
                connection.close();
            } catch (SQLException e1) {
                e1.printStackTrace();
            }
        }
    }

    @Override
    public boolean deleteAll() {
        Connection connection = databaseConnectionPool.getConnection();
        try (PreparedStatement pstm = connection.prepareStatement(DELETE_ALL)) {
            pstm.executeUpdate();
            return true;
        } catch (SQLException e) {
            e.printStackTrace();
            return false;
        } finally {
            try {
                connection.close();
            } catch (SQLException e1) {
                e1.printStackTrace();
            }
        }
    }

    @Override
    public List<BriefAccount> findBriefAccountsOfUser(Integer userId) {
        Connection connection = databaseConnectionPool.getConnection();
        try (PreparedStatement pstm = connection.prepareStatement(FIND_ACCOUNT_OF_USER)) {
            pstm.setInt(1, userId);

            ResultSet resultSet = pstm.executeQuery();
            List<BriefAccount> accounts = new ArrayList<BriefAccount>();
            while (resultSet.next()) {
                BriefAccount account = BriefAccount.builder()
                        .id(resultSet.getLong("id"))
                        .balance(resultSet.getDouble("balance"))
                        .accountName(resultSet.getString("account_name"))
                        .bank(Bank.builder()
                                .id(resultSet.getInt("bank_id"))
                                .name(resultSet.getString("bank_name"))
                                .bankCode(resultSet.getString("bank_code"))
                                .build())
                        .build();
                accounts.add(account);
            }
            return accounts;
        } catch (SQLException e) {
            e.printStackTrace();
            return null;
        } finally {
            try {
                connection.close();
            } catch (SQLException e1) {
                e1.printStackTrace();
            }
        }
    }

    @Override
    public List<AccountWithTransactionHistory> findAccountWithTransactionHistorysOfUser(Integer userId) {
        Connection connection = databaseConnectionPool.getConnection();
        List<AccountWithTransactionHistory> accounts = new ArrayList<AccountWithTransactionHistory>();
        try (PreparedStatement pstm = connection.prepareStatement(FIND_ACCOUNT_WITH_TRANSACTION_HISTORY_OF_USER)) {
            pstm.setInt(1, userId);

            ResultSet resultSet = pstm.executeQuery();

            while (resultSet.next()) {
                AccountWithTransactionHistory account = AccountWithTransactionHistory.builder()
                        .id(resultSet.getLong("id"))
                        .accountNumber(resultSet.getString("account_number"))
                        .balance(resultSet.getDouble("balance"))
                        .accountName(resultSet.getString("account_name"))
                        .accountType(AccountType.valueOf(resultSet.getString("account_type")))
                        .interestRatePerAnnum(resultSet.getDouble("interest_rate_per_annum"))
                        .bank(Bank.builder()
                                .id(resultSet.getInt("bank_id"))
                                .name(resultSet.getString("bank_name"))
                                .bankCode(resultSet.getString("bank_code"))
                                .build())
                        .build();
                accounts.add(account);
            }
        } catch (SQLException e) {
            try {
                connection.close();
            } catch (SQLException e1) {
                e1.printStackTrace();
            }
            e.printStackTrace();
            return null;
        }

        for (AccountWithTransactionHistory account : accounts) {
            account.setRecentTransactionHistoryDetails(
                    transactionHistoryRepository.findRecentTransactionHistoryDetailOfAccount(account.getId()));
        }
        try {
            connection.close();
        } catch (SQLException e1) {
            e1.printStackTrace();
        }

        return accounts;
    }

}

File: be/src/main/kotlin/sg/flow/repositories/account/AccountRepositoryImpl.kt
package sg.flow.repositories.account

import java.time.LocalDateTime
import kotlinx.coroutines.flow.toList
import kotlinx.coroutines.reactive.asFlow
import kotlinx.coroutines.reactive.awaitFirstOrNull
import org.springframework.r2dbc.core.DatabaseClient
import org.springframework.r2dbc.core.awaitRowsUpdated
import org.springframework.stereotype.Repository
import sg.flow.entities.Account
import sg.flow.entities.Bank
import sg.flow.entities.User
import sg.flow.entities.utils.AccountType
import sg.flow.models.account.AccountWithTransactionHistory
import sg.flow.models.account.BriefAccount
import sg.flow.repositories.transactionHistory.TransactionHistoryRepository
import sg.flow.repositories.utils.AccountQueryStore

@Repository
class AccountRepositoryImpl(
        private val databaseClient: DatabaseClient,
        private val transactionHistoryRepository: TransactionHistoryRepository,
) : AccountRepository {

        override suspend fun save(entity: Account): Account {
                val hasId = entity.id != null
                val queryString =
                        if (hasId) AccountQueryStore.SAVE_ACCOUNT_WITH_ID
                        else AccountQueryStore.SAVE_ACCOUNT

                if (hasId) {
                        databaseClient
                                .sql(queryString)
                                .bind(0, entity.id!!)
                                .bind(1, entity.accountNumber)
                                .bind(2, entity.bank.id!!)
                                .bind(3, entity.owner?.id!!)
                                .bind(4, entity.balance)
                                .bind(5, entity.accountName)
                                .bind(6, entity.accountType.toString())
                                .bind(7, entity.lastUpdated)
                                .fetch()
                                .awaitRowsUpdated()
                } else {
                        databaseClient
                                .sql(queryString)
                                .bind(0, entity.accountNumber)
                                .bind(1, entity.bank.id!!)
                                .bind(2, entity.owner?.id!!)
                                .bind(3, entity.balance)
                                .bind(4, entity.accountName)
                                .bind(5, entity.accountType.toString())
                                .bind(6, entity.lastUpdated)
                                .fetch()
                                .awaitRowsUpdated()
                }

                return entity
        }

        override suspend fun findById(id: Long): Account? {
                return databaseClient
                        .sql(AccountQueryStore.FIND_ACCOUNT_BY_ID)
                        .bind(0, id)
                        .map { row ->
                                Account(
                                        id = row.get("id", Long::class.java)!!,
                                        accountNumber =
                                                row.get("account_number", String::class.java)!!,
                                        balance = row.get("balance", Double::class.java)!!,
                                        accountName = row.get("account_name", String::class.java)!!,
                                        accountType =
                                                AccountType.valueOf(
                                                        row.get(
                                                                "account_type",
                                                                String::class.java
                                                        )!!
                                                ),
                                        lastUpdated =
                                                row.get(
                                                        "last_updated",
                                                        LocalDateTime::class.java
                                                )!!,
                                        interestRatePerAnnum =
                                                row.get(
                                                        "interest_rate_per_annum",
                                                        Double::class.java
                                                )!!,
                                        bank =
                                                Bank(
                                                        id = row.get("bank_id", Int::class.java)!!,
                                                        name =
                                                                row.get(
                                                                        "bank_name",
                                                                        String::class.java
                                                                )!!,
                                                        bankCode =
                                                                row.get(
                                                                        "bank_code",
                                                                        String::class.java
                                                                )!!
                                                ),
                                        finverseId = row.get(
                                                "finverse_id",
                                                String::class.java
                                        ),
                                        owner =
                                                User(
                                                        id = row.get("user_id", Int::class.java)!!,
                                                        name =
                                                                row.get(
                                                                        "name",
                                                                        String::class.java
                                                                )!!,
                                                        email =
                                                                row.get(
                                                                        "email",
                                                                        String::class.java
                                                                )!!,
                                                        identificationNumber =
                                                                row.get(
                                                                        "identification_number",
                                                                        String::class.java
                                                                )!!,
                                                        phoneNumber =
                                                                row.get(
                                                                        "phone_number",
                                                                        String::class.java
                                                                )!!,
                                                        dateOfBirth =
                                                                row.get(
                                                                        "date_of_birth",
                                                                        java.time.LocalDate::class
                                                                                .java
                                                                )!!,
                                                        address =
                                                                row.get(
                                                                        "address",
                                                                        String::class.java
                                                                )
                                                                        ?: "",
                                                        settingJson =
                                                                row.get(
                                                                        "setting_json",
                                                                        String::class.java
                                                                )
                                                                        ?: "{}"
                                                )
                                )
                        }
                        .one()
                        .awaitFirstOrNull()
        }

        override suspend fun deleteAll(): Boolean {
                return try {
                        databaseClient
                                .sql(AccountQueryStore.DELETE_ALL_ACCOUNTS)
                                .fetch()
                                .awaitRowsUpdated()
                        true
                } catch (e: Exception) {
                        e.printStackTrace()
                        false
                }
        }

        override suspend fun findBriefAccountsOfUser(userId: Int): List<BriefAccount> {
                var temp =
                        runCatching {
                                databaseClient
                                        .sql(AccountQueryStore.FIND_BRIEF_ACCOUNT_OF_USER)
                                        .bind(0, userId)
                                        .map { row ->
                                                BriefAccount(
                                                        id = row.get("id", Long::class.java)!!,
                                                        balance =
                                                                row.get(
                                                                        "balance",
                                                                        Double::class.java
                                                                )!!,
                                                        accountName =
                                                                row.get(
                                                                        "account_name",
                                                                        String::class.java
                                                                )!!,
                                                        bank =
                                                                Bank(
                                                                        id =
                                                                                row.get(
                                                                                        "bank_id",
                                                                                        Int::class
                                                                                                .java
                                                                                )!!,
                                                                        name =
                                                                                row.get(
                                                                                        "bank_name",
                                                                                        String::class
                                                                                                .java
                                                                                )!!,
                                                                        bankCode =
                                                                                row.get(
                                                                                        "bank_code",
                                                                                        String::class
                                                                                                .java
                                                                                )!!
                                                                )
                                                )
                                        }
                                        .all()
                                        .asFlow()
                                        .toList()
                        }
                                .onFailure { e ->
                                        e.printStackTrace()
                                        println("Error fetching brief accounts for userId: $userId")
                                }
                                .getOrElse { emptyList<BriefAccount>() }
                return temp
        }

        override suspend fun findAccountWithTransactionHistorysOfUser(
                userId: Int
        ): List<AccountWithTransactionHistory> {
                val accounts =
                        databaseClient
                                .sql(
                                        AccountQueryStore
                                                .FIND_ACCOUNT_WITH_TRANSACTION_HISTORY_OF_USER
                                )
                                .bind(0, userId)
                                .map { row ->
                                        AccountWithTransactionHistory(
                                                id = row.get("id", Long::class.java)!!,
                                                accountNumber =
                                                        row.get(
                                                                "account_number",
                                                                String::class.java
                                                        )!!,
                                                balance = row.get("balance", Double::class.java)!!,
                                                accountName =
                                                        row.get(
                                                                "account_name",
                                                                String::class.java
                                                        )!!,
                                                accountType =
                                                        AccountType.valueOf(
                                                                row.get(
                                                                        "account_type",
                                                                        String::class.java
                                                                )!!
                                                        ),
                                                interestRatePerAnnum =
                                                        row.get(
                                                                "interest_rate_per_annum",
                                                                Double::class.java
                                                        )!!,
                                                bank =
                                                        Bank(
                                                                id =
                                                                        row.get(
                                                                                "bank_id",
                                                                                Int::class.java
                                                                        )!!,
                                                                name =
                                                                        row.get(
                                                                                "bank_name",
                                                                                String::class.java
                                                                        )!!,
                                                                bankCode =
                                                                        row.get(
                                                                                "bank_code",
                                                                                String::class.java
                                                                        )!!
                                                        )
                                        )
                                }
                                .all()
                                .asFlow()
                                .toList()

                // Populate transaction history for each account
                accounts.forEach { account ->
                        account.recentTransactionHistoryDetails =
                                transactionHistoryRepository
                                        .findRecentTransactionHistoryDetailOfAccount(account.id)
                }

                return accounts
        }
}
